src/macro.jl

```diff
@@ -0,0 +1,14 @@
+"""
+    Adapt.@adapt_structure T
+
+Define a method `adapt_structure(to, obj::T)` which calls `adapt_structure` on each field
+of `obj` and constructs a new instance of `T` using the default constuctor.
+"""
+macro adapt_structure(T)
+    names = fieldnames(Core.eval(__module__, T))
+    quote
+        function Adapt.adapt_structure(to, obj::$(esc(T)))
+            $(esc(T))($([:(Adapt.adapt_structure(to, obj.$name)) for name in names]...))
```

Comment:
The existance of the default constructor is not guaranteed.

This might be fine (this would just error), or one could be more adventurous and do `Expr(:new, T, ...)`.

Reply:
Yeah, it makes some assumptions, but would probably work well for most cases.

What does `Expr(:new` do? I saw that in StructArrays.jl?

Reply:
It's `new()` i.e. it let's you pretend you are a inner constructor.

Reply:
Will that work? typically this will return an object of a different type.

Reply:
Let's be conservative for now, but yeah it should work.